### PR TITLE
Refactor `ToggleButtons` (remove `RawMaterialButton`)

### LINKED ISSUE #97302: ToggleButtons with  MaterialTapTargetSize.padded and height constrained to 32
This issue corresponds to Google internal issue b/204226713. **It does not need additional triage.**

The ToggleButtons height has been constrained to 32 but the input area has been padded, per MaterialTapTargetSIze.padded, to 48.  The Toggle button has been wrapped in a green Container to highlight the input padding.

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

class Home extends StatefulWidget {
  const Home({ Key? key }) : super(key: key);

  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  int selectedTabIndex = 0;

  @override
  Widget build(BuildContext context) {
    final ColorScheme colorScheme = Theme.of(context).colorScheme;
    return Scaffold(
      body: Center(
        child: Container(
          color: Colors.green.withOpacity(0.15),
          child: ToggleButtons(
            tapTargetSize: MaterialTapTargetSize.padded,
            isSelected: [
              selectedTabIndex == 0,
              selectedTabIndex == 1,
            ],
            onPressed: (int index) {
              setState(() { selectedTabIndex = index; });
            },
            borderRadius: const BorderRadius.all(Radius.circular(8)),
            selectedBorderColor: colorScheme.primary,
            selectedColor: colorScheme.primary,
            constraints: const BoxConstraints.tightFor(
              width: 86,
              height: 32,
            ),
            children: [
              Text('Locks'),
              Text('Unlocks'),
            ],
          ),
        ),
      ),
    );
  }
}
void main() {
  runApp(const MaterialApp(home: Home()));
}
```
</details>

### LINKED ISSUE #99085: Refactor ToggleButtons
ToggleButtons use RawMaterialButton which is scheduled to be deprecated and removed (eventually). 
We need to refactor them to use the new button universe, either TextButton, OutlinedButton or ElevatedButton.

cc @TahaTesser and @HansMuller 

Completing this will likely resolve https://github.com/flutter/flutter/issues/97302 as well.

### PR DESCRIPTION
**Pull request by TahaTesser**
fixes https://github.com/flutter/flutter/issues/99085
fixes https://github.com/flutter/flutter/issues/97302

<details> 
<summary>minimal code sample</summary> 

```dart
import 'package:flutter/material.dart';

const List<Widget> icons = <Widget>[
  Icon(Icons.ac_unit),
  Icon(Icons.call),
  Icon(Icons.cake),
];

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Material App',
      home: ToggleButtonsSample(),
    );
  }
}

class ToggleButtonsSample extends StatelessWidget {
  ToggleButtonsSample({Key? key}) : super(key: key);
  final ValueNotifier<int> _toggleValue = ValueNotifier<int>(0);

  @override
  Widget build(BuildContext context) {
    final ColorScheme colorScheme = Theme.of(context).colorScheme;

    return Scaffold(
      appBar: AppBar(
        title: const Text('ToggleButtons Sample'),
      ),
      body: Center(
        child: ValueListenableBuilder(
          valueListenable: _toggleValue,
          builder: (BuildContext context, int value, Widget? child) {
            return ToggleButtons(
              onPressed: (int newIndex) => _toggleValue.value = newIndex,
              borderRadius: const BorderRadius.all(Radius.circular(8)),
              selectedBorderColor: colorScheme.primary,
              selectedColor: colorScheme.primary,
              isSelected: <bool>[
                _toggleValue.value == 0,
                _toggleValue.value == 1,
                _toggleValue.value == 2,
              ],
              children: icons,
            );
          },
        ),
      ),
    );
  }
}
``` 
	
</details>

### Description
Here this PR refactors `ToggleButtons` and use `ButtonStyleButton` to style `ToggleButtons` 

highlight color not being translated to ElevatedButton (`ButtonStyleButton` sets highlight color to be [transparent](https://github.com/flutter/flutter/blob/6e2d31459d08d27e6076b087244b8bc7811687cf/packages/flutter/lib/src/material/button_style_button.dart#L368)).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by Piinks — COMMENTED** [review]
NICE - all of the existing golden file tests pass! 🎉  (There are only two, but still, that's great)

Comparing this to the same refactor you are doing for FAB (https://github.com/flutter/flutter/pull/99753)

- FAB is not becoming a `ButtonStyleButton`, but here it looks like `ToggleButtons` are.
- FAB is not going to have a `ButtonStyle style` parameter like here, instead deferring to its numerated properties and inherited theme to create a ButtonStyle internally. ToggleButtons here allow for style to be set OR individual properties + theming.

I wonder if the FAB approach is cleaner, the style property here on ToggleButtons might not be necessary, but again I will defer to @HansMuller on general design since he wrote the model. 

All of this work in FAB & Toggle is really great progress though. Thanks!

**Comment by HansMuller** [comment]
We should make sure that this PR also fixes https://github.com/flutter/flutter/issues/97302

**Comment by HansMuller** [comment]
I agree [with Kate](https://github.com/flutter/flutter/pull/99493#pullrequestreview-906533911), this PR shouldn't include the extra step of adding a ButtonStyle `style` parameter to ToggleButtons; we should leave that for a separate proposal/PR.

**Comment by TahaTesser** [comment]
> We should make sure that this PR also fixes https://github.com/flutter/flutter/issues/97302

This is now handled by `TextButton` widget in this refactor, can you please take a look

### Mobile

Screenshot_1647249245

### Desktop (Web)
Screenshot 2022-03-14 111522

**Comment by HansMuller** [comment]
@TahaTesser - if this PR fixes #97302 it should say so in the PR's description and it should include a regression test (marked as a regression test for #97302 with a comment).

**Comment by HansMuller** [comment]
@TahaTesser - the screenshots in https://github.com/flutter/flutter/pull/99493#issuecomment-1067042366 imply that the background color for the ToggleButtons isn't clipped to the ToggleButtons' shape?

**Comment by TahaTesser** [comment]
@HansMuller 
Thanks so much for your time,

> @TahaTesser - the screenshots in https://github.com/flutter/flutter/pull/99493#issuecomment-1067042366 imply that the background color for the ToggleButtons isn't clipped to the ToggleButtons' shape?

This is just a parent container with color from https://github.com/flutter/flutter/issues/9730. 

ClipRRect still clipping the toggle buttons in this refactor
https://github.com/flutter/flutter/pull/99493/files#diff-153bdbb7a6739dd071bbb7b8230e3884dcb6cedbc21a78ef82c0e6a8166bb220R731

**Comment by HansMuller** [comment]
@TahaTesser - Good!

**Comment by TahaTesser** [comment]
@HansMuller 
Added a regression test with a link to the issue

I am currently investigating this behavior, just noticed this, this seems to be only happening in the vertical direction.
| Preview | 
| --------------- | 
| 	|

**Comment by TahaTesser** [comment]
@HansMuller 
During refactoring noticed, there is no interactive example that showcases different configurations for `ToggleButtons`

Filed https://github.com/flutter/flutter/pull/100124, this will land after this PR (tester finders look for `TextButton` instead of `RawMaterialButton`).

**Review by HansMuller — COMMENTED** [review]
This is looking good however I don't think the style parameter belongs in this PR; see https://github.com/flutter/flutter/pull/99493#issuecomment-1064643880

**Comment by TahaTesser** [comment]
> This is looking good however I don't think the style parameter belongs in this PR; see https://github.com/flutter/flutter/pull/99493#issuecomment-1064643880

My apologies I missed removing that parameter, done.

File: packages/flutter/lib/src/material/toggle_buttons.dart
Comment (TahaTesser) [reply packages/flutter/lib/src/material/toggle_buttons.dart]: Thanks! Removed nullable except `mouseCursor`  since it is using a direct nullable widget parameter.

File: packages/flutter/lib/src/material/toggle_buttons.dart
Comment (TahaTesser) [reply packages/flutter/lib/src/material/toggle_buttons.dart]: I merged the two and it is provided directly to the` ButtonStyle` now since the conditioning is inside `_ResolveFillColor` now.

File: packages/flutter/lib/src/material/toggle_buttons.dart
Comment (TahaTesser) [reply packages/flutter/lib/src/material/toggle_buttons.dart]: Removed these redundant style properties.

**Review by Piinks — COMMENTED** [review]
@TahaTesser is this ready for another review? Or is it waiting on additional changes like the FAB one?

**Comment by TahaTesser** [comment]
> @TahaTesser is this ready for another review? Or is it waiting on additional changes like the FAB one?

Hey @Piinks 
This needs another review. 

This `Diagnosticable` issue from FAB one doesn't affect this PR, all tests are passing

**Comment by HansMuller** [comment]
@TahaTesser - I haven't tested or even compiled the code I've suggested for the _ForegroundColor and _BackgroundColor. If they're difficult let me know and I'll fix them.

**Comment by TahaTesser** [comment]
> @TahaTesser - I haven't tested or even compiled the code I've suggested for the _ForegroundColor and _BackgroundColor. If they're difficult let me know and I'll fix them.

This is cleaner, thanks! 
I made slight adjustments and added these classes including the local ColorScheme variable as suggested and removed the redundant code due to these classes.

File: packages/flutter/lib/src/material/toggle_buttons.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/toggle_buttons.dart:697]: If you define a isSelected variable above, then these constructors and their positional parameters should fit neatly on one line. The same `onPressed != null && isSelected[index]` appears in the overlayColor value too; definitely wouldn't hurt to factor it out.

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/material/toggle_buttons.dart
Comment (TahaTesser) [reply packages/flutter/lib/src/material/toggle_buttons.dart]: Great idea, I added a local variable and made those into a single line 

Also added regression for https://github.com/flutter/flutter/issues/97302 (I added it earlier too but somehow missed it during buttonstyle revert).  This test would be able to close the issue :)

**Comment by Piinks** [comment]
I just got an alert that this failed 1 internal test, but the test itself needs to be fixed simultaneously with this change landing. It is possible to do, let's just hold on landing it until I have the fix ready to land. I'll ping back once it's ready.

**Comment by Piinks** [comment]
Update: The fix is out for review, awaiting approval.

**Comment by Piinks** [comment]
@TahaTesser, can you update your branch with the latest from master? The google testing shard is stuck. I have the fix ready, so once it is rebased we can land it.

**Comment by math1man** [comment]
Hey, I think this change is a incorrect. https://github.com/flutter/flutter/pull/99493#issuecomment-1067042366 does not look like the correct toggle button styling on Mobile.

**Comment by HansMuller** [comment]
@math1man - can you explain what aspect of the component doesn't look correct?

**Comment by math1man** [comment]
It's too tall.

**Comment by HansMuller** [comment]
@TahaTesser - can you investigate? AFAICT we have not changed the geometry of ToggleButtons, however perhaps we've always had this wrong. I don't believe https://m3.material.io/ covers it (yet).

**Comment by Piinks** [comment]
The failure appears related to https://github.com/flutter/flutter/issues/97302, which this closed. Rather than allow for a height of 32, it seems like this enforced a height of 48 - IIRC that is the minimum required size for interactive elements. To reproduce this, the ToggleButtons.constraints were set with a min/max height of 32, but did not adhere to that.

**Comment by TahaTesser** [comment]
@HansMuller 

I also found this in the Material Design components app, I will look into and test it in an Android native app.

| Preview | 
| --------------- | 
| 	|


Perhaps there should Android equivalent of https://github.com/flutter/platform_tests/tree/master/ios_widget_catalog_compare

**Comment by TahaTesser** [comment]
@Piinks 
Thank you for reverting

**Comment by HansMuller** [comment]
If a PR is created that relands the ToggleButton refactor, we'll need some additional regression tests to verify that the buttons handle input within 48 vertical without the button's outline occupying 48.

**Pull request merged**
